### PR TITLE
DOT importer using antlr

### LINKED ISSUE #206: DOTImporter implementation not suitable for large graphs
Hi to all,

I am using JGrapht for a university project in which I work with large network data which contains about 8000 nodes. I am looking for a way to serialize the generated graph in order to re-use it. 

I have seen that you support the default Java serialization mechanism and the DOT serialization format.  I have tried the latter because for my application could be interesting to plot the graph too. Unfortunately, if you work with large graphs your implementation of the DOTImporter it is not appropriate because the _read_ method expects the whole file content in input as a string. In my opinion the DOTImporter should provide a _read_ method which reads from a _Reader_ or _BufferedReader_, in order to incrementally build the graph. 

Do you think that this solution is possible or am I missing something?

Thank you in advance for your answer.

Alessandro

### PR DESCRIPTION
**Pull request by d-michail**
A from scratch implementation of the DOTImporter using antlr. 

- Handles the full specification including XML string identifiers, etc.
- Added dependency on commons-lang3 for escape-unescape.
- Updated antlr version to 4.6.
- Resolves #206 and makes #318 obsolete.
- The old implementation is still available as deprecated in org.jgrapht.ext.

### CODE REVIEW

**Comment by jkinable** [comment]
I got a bit confused: what's the difference between DOTImporterTest.java and DOTImporter2Test.java?

**Comment by d-michail** [comment]
They are testing more advanced functionality, thus I kept them in a separate class. Maybe renaming the class will make things more intuitive?

**Comment by jkinable** [comment]
An impressive piece of coding! Great job!

I'll not pretend that I've read everything in detail... it would be pretty much impossible for me to check for errors unless I spend significantly more time studying lexers/parsers. The tests look solid though. 

I would recommend naming the test classes DOTImporterTests1.java and DOTImporterTests2.java or something, and maybe add a comment in DOTImporterTests2 mentioning that this is a continuation of DOTImporterTests1. The current names (DOTImporterTest and EnhancedDOTImporterTest) may cause issues in the future since they don't appear consecutively when the files are sorted alphabetically. It's easy to overlook one or the other.

@jsichi do you want to look at this, or shall we go ahead and merge this one. Any remaining implementation errors should surface relatively quickly when people start using this importer/exporter.

File: jgrapht-ext/src/test/java/org/jgrapht/ext/DOTImporterTest.java [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+
+import java.io.*;
+import java.util.*;
+
+import org.jgrapht.*;
+import org.jgrapht.graph.*;
+
+import junit.framework.*;
+
+public class DOTImporterTest
```
Comment (jkinable) [inline jgrapht-ext/src/test/java/org/jgrapht/ext/DOTImporterTest.java]: Should this class be deprecated?

**Comment by jsichi** [comment]
ship it

**Comment by jkinable** [comment]
Alright, done. :)

**Pull request merged**